cpp/src/reductions/group_std.cu

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "group_reductions.hpp"
+
+#include <bitmask/legacy/bit_mask.cuh>
+#include <cudf/utilities/legacy/type_dispatcher.hpp>
+
+#include <thrust/reduce.h>
+#include <thrust/iterator/discard_iterator.h>
+
+namespace {
+
+template <typename T>
+void print(rmm::device_vector<T> const& d_vec, std::string label = "") {
+  thrust::host_vector<T> h_vec = d_vec;
+  printf("%s \t", label.c_str());
+  for (auto &&i : h_vec)  std::cout << i << " ";
+  printf("\n");
+}
+
+template <typename T>
+void print(gdf_column const& col, std::string label = "") {
+  auto col_data = reinterpret_cast<T*>(col.data);
+  auto d_vec = rmm::device_vector<T>(col_data, col_data+col.size);
+  print(d_vec, label);
+}
+
+struct var_functor {
+  template <typename T>
+  std::enable_if_t<std::is_arithmetic<T>::value, void >
+  operator()(gdf_column const& values,
+             rmm::device_vector<gdf_size_type> const& group_labels,
+             rmm::device_vector<gdf_size_type> const& group_sizes,
+             gdf_column * result,
+             gdf_size_type ddof,
+             cudaStream_t stream)
+  {
+    print(group_labels, "labels");
+    print(group_sizes, "sizes");
+    print<T>(values, "values");
+
+    auto values_data = static_cast<const T*>(values.data);
+    auto result_data = static_cast<double *>(result->data);
+    auto values_valid = reinterpret_cast<const bit_mask::bit_mask_t*>(values.valid);
+    auto result_valid = reinterpret_cast<bit_mask::bit_mask_t*>(result->valid);
+    const gdf_size_type* d_group_labels = group_labels.data().get();
+    const gdf_size_type* d_group_sizes = group_sizes.data().get();
+    
+    // Calculate sum
+    // TODO: replace with mean function call when that gets an internal API
+    rmm::device_vector<T> sums(values.size);
+
+    thrust::reduce_by_key(rmm::exec_policy(stream)->on(stream),
+                          group_labels.begin(), group_labels.end(),
+                          thrust::make_transform_iterator(
```

Comment:
It does work but as I've seen, it increases compile time and the amount of code. The latter is because nullable and non-nullable iterators are of different types and any thrust call that uses it, needs to be instantiated twice. The current approach doesn't really add to the run time.

Reply:
@kovaltan , Correct me if I'm wrong but we would still need two `reduce_by_key` operations, right? The second reduction requires the mean to be computed before it is called. With that new iterator, we **can** reorganize the computations. Note that the sum of squares needed is `(x - mean)^2` and not `x^2`.

> How about exclude nulls at pre-computed sort?

We **do** remove null keys at pre-computed sort. We also do it for values but the nulls end up in the end of each group. 

```
values        = {5, 6, 7, 8, 9, 10, 11};
val_nulls     = {1, 1, 1, 1, 1, 0, 1};
group_ids     = {0, 1, 1, 2, 2, 2, 3};
key_nulls     = {1, 1, 0, 0, 1, 1, 1};
valid_counts  = 5;  // valid_count_of(nulls);

sorted_keys   = {0,   1,   2, 2,    3}; // {1,2} are removed
sorted_values = {5,   6,   9, 10,   11};
                              ^ Null but followed by non-null in next group
```

This limitation is because of the use of a [helper object](https://github.com/rapidsai/cudf/blob/branch-0.10/cpp/src/quantiles/groupby.hpp) being used for sorting and segmenting the value column. That object memoizes all the building blocks for aggregations that are agnostic to the value column, so that we can reuse those vectors for sorting and aggregating other columns. Ideally we should've removed nulls from the value columns during sorting but that would require computing the `group_labels` (which is just a stand-in for key table) for each column. What we have now is an optimization at a small inconvenience to the developer who's writing aggregations.

Reply:
@devavret 
For `var` and `std`, we can also use x^2 instead of (x - mean)^2, though there is a risk regarding the accuracy of the precision.
Please look this I used for column var/std:
https://github.com/rapidsai/cudf/blob/625a6088d9ecce7a9ca70bba3afe0ddb92102164/cpp/src/reductions/reduction_operators.cuh#L196-L203


Aha, both of group id and value should have each nulls, the sorted keys should be shared with other columns in the table. That makes sense. Then, I have another question, where does the count of each group for selected column?
The count for each group for the column looks coming from input group_sizes, is it shared with other columns? How is it computed? Is there any chance to compute sorted keys when calculate the group_sizes for each column?
Btw, my example also shows how to compute count, sums and sum of squared value in a single call.